add OpenShift installation doc (#563)
Kuadrant/kuadrant-operator: doc/install/install-openshift.md

> # Install Kuadrant on an OpenShift Cluster
> 
> Note these steps need to be taken on each cluster you want to use Kuadrant on.

Reviewer: ```suggestion
# Install Kuadrant on an OpenShift cluster

NOTE: You must perform these steps on each cluster that you want to use Kuadrant on.
```

## Prerequisites

- OpenShift Container Platform 4.14.x or later with community Operator catalog available
- AWS account with Route 53 and zone 
- Accessible Redis Instance

> ## Setup Environment
> 

Reviewer: ```suggestion
## Set up your environment

```

```bash
export AWS_ACCESS_KEY_ID=xxxxxxx # Key ID from AWS with Route 53 access
export AWS_SECRET_ACCESS_KEY=xxxxxxx # Access Key from AWS with Route 53 access
export REDIS_URL=redis://user:xxxxxx@some-redis.com:10340 # A Redis cluster URL
```

> ## Installing the dependencies
> 
> Kuadrant integrates with Istio as a Gateway API provider. Before you can try Kuadrant, we need to setup an Istio based Gateway API provider. For this we will use the `Sail` operator.

Reviewer: ```suggestion
## Install the dependencies

Kuadrant integrates with Istio as a Gateway API provider. Before you can use Kuadrant, you must set up an Istio-based Gateway API provider. For this step, you will use the Sail Operator.
```

### Install v1 of Gateway API:

```bash
kubectl apply -f https://github.com/kubernetes-sigs/gateway-api/releases/download/v1.0.0/standard-install.yaml
```

### Install and Configure Istio via the Sail Operator

To install, run the following:

```bash
kubectl create ns istio-system
```

```bash
kubectl  apply -f - <<EOF
kind: OperatorGroup
apiVersion: operators.coreos.com/v1
metadata:
  name: sail
  namespace: istio-system
spec: 
  upgradeStrategy: Default  
---  
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: sailoperator
  namespace: istio-system
spec:
  channel: 3.0-dp1
  installPlanApproval: Automatic
  name: sailoperator
  source: community-operators
  sourceNamespace: openshift-marketplace
EOF
```

To check the status of the install, you can run:

```bash
kubectl get installplan -n istio-system -o=jsonpath='{.items[0].status.phase}'
```

> Once ready it will be marked `complete` while installing it will be marked `installing`.

Reviewer: ```suggestion
When ready, the status will change from `installing` to `complete`.
```

#### Configure Istio

```bash
kubectl apply -f - <<EOF
apiVersion: operator.istio.io/v1alpha1
kind: Istio
metadata:
  name: default
spec:
  version: v1.21.0
  namespace: istio-system
  # Disable autoscaling to reduce dev resources
  values:
    pilot:
      autoscaleEnabled: false
EOF
```

Wait for Istio to be ready:

```bash
kubectl wait istio/default -n istio-system --for="condition=Ready=true"
```

> ### (Recommended) Thanos and Observability Stack
> 
> Kuadrant provides a set of sample dashboards that use the known metrics exported by Kuadrant and Gateway components to provide insight into the different areas of your APIs and Gateways. While this is not essential, it is recommended that you set up an observability stack. Below are links to the OpenShift docs on this and also a link to help with the setup of Thanos for metrics storage.

Reviewer: ```suggestion
### Best practices for metrics and observability

Kuadrant provides a set of sample dashboards that use known metrics exported by Kuadrant and Gateway components to provide insight into different areas of your APIs and Gateways. While not essential, it is best to set up an observability stack. This section provides links to OpenShift and Thanos documentation on configuring monitoring and metrics storage.
```

OpenShift supports a user facing monitoring stack. This can be cofigured and setup this documentation:

> https://docs.openshift.com/container-platform/4.14/observability/monitoring/configuring-the-monitoring-stack.html

Reviewer: ```suggestion
https://docs.openshift.com/container-platform/latest/observability/monitoring/configuring-the-monitoring-stack.html
```

> https://docs.openshift.com/container-platform/4.14/observability/monitoring/configuring-the-monitoring-stack.html

Reviewer: ```suggestion
OpenShift supports a user-facing monitoring stack, which you can configure by using the steps in the [OpenShift monitoring documentation](https://docs.openshift.com/container-platform/latest/observability/monitoring/configuring-the-monitoring-stack.html).
```

If you have user workload monitoring enabled. We Recommend configuring  remote write to a central storage system such as Thanos: 

> - [Remote Write Config](https://docs.openshift.com/container-platform/4.14/observability/monitoring/configuring-the-monitoring-stack.html#configuring_remote_write_storage_configuring-the-monitoring-stack)

Reviewer: ```suggestion
- [Remote Write Config](https://docs.openshift.com/container-platform/latest/observability/monitoring/configuring-the-monitoring-stack.html#configuring_remote_write_storage_configuring-the-monitoring-stack)
```

- [Kube Thanos](https://github.com/thanos-io/kube-thanos)

> - [Kuadrant Examples](https://docs.kuadrant.io/kuadrant-operator/doc/observability/examples/)

Reviewer: ```suggestion
If you have user workload monitoring enabled, it is best to configure remote writing to a central storage system such as Thanos:
- [OpenShift remote write configuration](https://docs.openshift.com/container-platform/latest/observability/monitoring/configuring-the-monitoring-stack.html#configuring_remote_write_storage_configuring-the-monitoring-stack)

- [Kube Thanos](https://github.com/thanos-io/kube-thanos)

- [Kuadrant example dashboards and alerts](https://docs.kuadrant.io/kuadrant-operator/doc/observability/examples/)
```

### Install Kuadrant

To install Kuadrant, we will use the Kuadrant Operator. Prior to installing, we will set up some secrets that we will use later:

```bash
kubectl create ns kuadrant-system
```

AWS Route 53 credentials for TLS verification:

```bash
kubectl -n kuadrant-system create secret generic aws-credentials \
  --type=kuadrant.io/aws \
  --from-literal=AWS_ACCESS_KEY_ID=$AWS_ACCESS_KEY_ID \
  --from-literal=AWS_SECRET_ACCESS_KEY=$AWS_SECRET_ACCESS_KEY
```

> Redis credentials for shared (multi-cluster) counters support for Kuadrant's Limitador component:

Reviewer: ```suggestion
Redis credentials for shared multicluster counters for Kuadrant's Limitador component:
```

```bash
kubectl -n kuadrant-system create secret generic redis-config \
  --from-literal=URL=$REDIS_URL  
```  

```bash
kubectl create ns ingress-gateway
```

AWS Route 53 credentials for managing DNS records:

```bash
kubectl -n ingress-gateway create secret generic aws-credentials \
  --type=kuadrant.io/aws \
  --from-literal=AWS_ACCESS_KEY_ID=$AWS_ACCESS_KEY_ID \
  --from-literal=AWS_SECRET_ACCESS_KEY=$AWS_SECRET_ACCESS_KEY
```  

Finally, to install the Kuadrant Operator:

```bash
kubectl apply -f - <<EOF
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: kuadrant-operator
  namespace: kuadrant-system
spec:
  channel: preview
  installPlanApproval: Automatic
  name: kuadrant-operator
  source: community-operators
  sourceNamespace: openshift-operators
---
kind: OperatorGroup
apiVersion: operators.coreos.com/v1
metadata:
  name: kuadrant
  namespace: kuadrant-system
spec: 
  upgradeStrategy: Default 
EOF
```  

> Wait for Kuadrant operators to be installed:

Reviewer: ```suggestion
Wait for Kuadrant Operators to be installed:
```

```bash
kubectl get installplan -n kuadrant-system -o=jsonpath='{.items[0].status.phase}'
```

After some time, this should return `complete`.

#### Configure Kuadrant

```bash
kubectl apply -f - <<EOF
apiVersion: kuadrant.io/v1beta1
kind: Kuadrant
metadata:
  name: kuadrant
  namespace: kuadrant-system
spec:
  limitador:
    storage:
      redis-cached:
        configSecretRef:
          name: redis-config 
EOF          
```      

Wait for Kuadrant to be ready:

```bash
kubectl wait kuadrant/kuadrant --for="condition=Ready=true" -n kuadrant-system --timeout=300s
```

Kuadrant is now ready to use.

> Next up: [Secure, Protect and Connect on single or multiple clusters](../user-guides/secure-protect-connect-single-multi-cluster.md)

Reviewer: ```suggestion
## Next steps 
- [Secure, protect, and connect APIs on single or multiple clusters](../user-guides/secure-protect-connect-single-multi-cluster.md)
```

# Install Kuadrant on an OpenShift Cluster

Note these steps need to be taken on each cluster you want to use Kuadrant on.

> ## Pre Requisites
> 
> - OpenShift 4.14.x with community operator catalog available

Reviewer: ```suggestion
## Prerequisites

- OpenShift Container Platform 4.14.x or later with community Operator catalog available
```

Reviewer: QE test plan mentions OCP 4.15

# Install Kuadrant on an OpenShift Cluster

Note these steps need to be taken on each cluster you want to use Kuadrant on.

## Pre Requisites

- OpenShift 4.14.x with community operator catalog available
- AWS account with route 53 and zone 
- Accessible Redis Instance

## Setup Environment

```
export AWS_ACCESS_KEY_ID=xxxxxxx # the key id from AWS
export AWS_SECRET_ACCESS_KEY=xxxxxxx # the access key from AWS
export REDIS_URL=redis://user:xxxxxx@some-redis.com:10340

```

## Installing the dependencies

Kuadrant integrates with Istio as a Gateway API provider. Before trying Kuadrant, we need to setup an Istio based Gateway API provider. For this we will use the `Sail` operator. 

- Install v1 of Gateway API

```
kubectl apply -f https://github.com/kubernetes-sigs/gateway-api/releases/download/v1.0.0/standard-install.yaml
```

- ### Install and Configure Istio via the Sail Operator

#### Install
```
kubectl create ns istio-system
```

```
kubectl  apply -f - <<EOF
kind: OperatorGroup
apiVersion: operators.coreos.com/v1
metadata:
  annotations:

>     argocd.argoproj.io/sync-wave: "0"

Reviewer 2: Argo annotation should not be present

  name: sail
  namespace: istio-system
spec: 
  upgradeStrategy: Default  
---  
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: sailoperator
  namespace: istio-system
spec:
  channel: 3.0-dp1
  installPlanApproval: Automatic
  name: sailoperator
  source: community-operators
  sourceNamespace: openshift-marketplace
EOF
```

To check the status of the install you can run:

```
kubectl get installplan -n istio-system -o=jsonpath='{.items[0].status.phase}'
```

once ready it will be marked `complete` while installing it will be marked `installing`

#### Configure Istio

```
kubectl apply -f - <<EOF
apiVersion: operator.istio.io/v1alpha1
kind: Istio
metadata:
  annotations:

>     argocd.argoproj.io/sync-wave: "1"

Reviewer 2: Argo annotation should not be present

EOF
```

Wait for Istio to be ready

```
kubectl wait istio/default -n istio-system --for="condition=Ready=true"
```

### (Recommended) Thanos and Observability Stack

Kuadrant provides a set of sample dashboards that use the known metrics exported by kuadrant and gateway components to provide insight into the different areas of your APIs and Gateways. While this is not essential it is recommended that you set up an observability stack. Below are links to the OpenShift docs on this and also a link to help with the setup of Thanos for metrics storage.

OpenShift supports a user facing monitoring stack. This can be cofigured and setup via their documentation
https://docs.openshift.com/container-platform/4.14/observability/monitoring/configuring-the-monitoring-stack.html

- If you have user workload monitoring enabled. We Recommend configuring  remote write to a central storage system such as Thanos: 
- [Remote Write Config](https://docs.openshift.com/container-platform/4.14/observability/monitoring/configuring-the-monitoring-stack.html#configuring_remote_write_storage_configuring-the-monitoring-stack)

- [Kube Thanos](https://github.com/thanos-io/kube-thanos)

> - [Kuadrant Examples](https://docs.kuadrant.io/kuadrant-operator/doc/observability/examples/)

Reviewer 3: I'm thinking about enhancing the setup instructions for Thanos and the Observability Stack to help users get up and running more quickly. I agree that they can find everything important in the guides provided, but the idea is to help them get whole stuck up and running quickly.

> - [Kuadrant Examples](https://docs.kuadrant.io/kuadrant-operator/doc/observability/examples/)

Reviewer 3: Opening up a new repo is great, but this par only points to 4 other links, and its informative value is low. I am currently going through the documentation mentioned above and trying to get it to work.
We are documenting how to install Sail, so it does make sense to do the same for the Observability stack.

> - [Kuadrant Examples](https://docs.kuadrant.io/kuadrant-operator/doc/observability/examples/)

Author: No the observability stack is optional and far more complex to maintain. Istio as a gateway provider is a hard dependency

> - [Kuadrant Examples](https://docs.kuadrant.io/kuadrant-operator/doc/observability/examples/)

Reviewer 2: I would argue that everything we are showing is optional, but observability was one of the requirements for this release and a big feature. If the scope of this doc is to showcase everything important we should make more effort to help users get up to speed and to help them set up the feature. We do not need to copy paste openshift doc but we could link to specific sections that might be of particular interest for the simplest use case we are trying to show.

Also Grafana is not covered as part of openshift guide and should have some instruction how to install as well

> - [Kuadrant Examples](https://docs.kuadrant.io/kuadrant-operator/doc/observability/examples/)

Reviewer 4: > do we have a good ref on grafana install

maybe https://cloud.redhat.com/experts/o11y/ocp-grafana/ ?
It can be installed from operatorhub as well I believe

> enhancing the setup instructions for Thanos and the Observability Stack

To keep things easier for this version, and also because of known issues with aggregations from multiple clusters in dashboards, we could direct users to configure UWM in 1 cluster only (no remote write instructions). 
Then deploy anything needed to get UWM scraping from istio, envoy, gateway-api-state-metrics (via custom install of kube-state-metrics).
Finally, point to the new kuadrant dashboards doc for how to import dashboards.

I can help with pointing to the right resources for this if agreeable?

> - [Kuadrant Examples](https://docs.kuadrant.io/kuadrant-operator/doc/observability/examples/)

Author: @david-martin do you want to add PR to this branch with any changes you feel make sense

### Install Kuadrant

To install Kuadrant we will use the Kuadrant Operator. Before this lets set up some secrets we will need later.

```

kubectl create ns kuadrant-system

```

AWS Credential for TLS verification

```
kubectl -n kuadrant-system create secret generic aws-credentials \
  --type=kuadrant.io/aws \
  --from-literal=AWS_ACCESS_KEY_ID=$AWS_ACCESS_KEY_ID \
  --from-literal=AWS_SECRET_ACCESS_KEY=$AWS_SECRET_ACCESS_KEY
````  

Redis Credential for shared counters

```
kubectl -n kuadrant-system create secret generic redis-config \
  --from-literal=URL=$REDIS_URL  
```  

```
kubectl create ns ingress-gateway
```

AWS Credential for managing DNS records

```
kubectl -n ingress-gateway create secret generic aws-credentials \
  --type=kuadrant.io/aws \
  --from-literal=AWS_ACCESS_KEY_ID=$AWS_ACCESS_KEY_ID \
  --from-literal=AWS_SECRET_ACCESS_KEY=$AWS_SECRET_ACCESS_KEY
```  

To install the Kuadrant operator:      
```
kubectl apply -f - <<EOF
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: kuadrant-operator
  namespace: kuadrant-system
spec:
  channel: preview
  installPlanApproval: Automatic
  name: kuadrant-operator

>   source: kuadrant-operator-catalog

Reviewer 2: If we would like to direct users to use the released Kuadrant (from OperatorHub) we should use:
```
  source: community-operators
  sourceNamespace: openshift-operators
```

Author: good spot. This was left over from testing

  sourceNamespace: openshift-operators
---
kind: OperatorGroup
apiVersion: operators.coreos.com/v1
metadata:
  name: kuadrant
  namespace: kuadrant-system
spec: 
  upgradeStrategy: Default 
EOF
```  

Wait for kuadrant operators to be installed

```
kubectl get installplan -n kuadrant-system -o=jsonpath='{.items[0].status.phase}'
```

This should return `complete`

#### Configure Kuadrant

```
kubectl apply -f - <<EOF
apiVersion: kuadrant.io/v1beta1
kind: Kuadrant
metadata:
  annotations:

>     argocd.argoproj.io/sync-wave: "1"

Reviewer 2: Argo annotation should not be present

OpenShift supports a user facing monitoring stack. This can be cofigured and setup this documentation:

https://docs.openshift.com/container-platform/latest/observability/monitoring/configuring-the-monitoring-stack.html

If you have user workload monitoring enabled. We Recommend configuring remote write to a central storage system such as Thanos:

- [Remote Write Config](https://docs.openshift.com/container-platform/latest/observability/monitoring/configuring-the-monitoring-stack.html#configuring_remote_write_storage_configuring-the-monitoring-stack)
- [Kube Thanos](https://github.com/thanos-io/kube-thanos)

There are a set of [example dashboards and alerts](https://docs.kuadrant.io/kuadrant-operator/doc/observability/examples/) for observing Kuadrant functionality.
These dashboards and alerts make use of low level cpu, metrics and network metrics available from the user monitoring stack in Openshift. They also make use of resource state metrics from Gateway API and Kuadrant resources.
To scrape these additional metrics, you can install a kube-state-metrics instance, with a custom resource config:

```bash

> kubectl apply -f https://raw.githubusercontent.com/Kuadrant/kuadrant-operator/main/config/observability/openshift/kube-state-metrics.yaml

Reviewer 4: note the `main` in the url will not work until this PR lands as the file referenced is new in this pr.

If trying this out now, you can use this:

```sh
kubectl apply -f https://raw.githubusercontent.com/Kuadrant/kuadrant-operator/install-doc/config/observability/openshift/kube-state-metrics.yaml
```

```

```
kubectl  apply -f - <<EOF
kind: OperatorGroup
apiVersion: operators.coreos.com/v1
metadata:
  annotations:
    argocd.argoproj.io/sync-wave: "0"
  name: sail
  namespace: istio-system
spec: 
  upgradeStrategy: Default  
---  
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  name: sailoperator
  namespace: istio-system
spec:
  channel: 3.0-dp1
  installPlanApproval: Automatic
  name: sailoperator
  source: community-operators
  sourceNamespace: openshift-marketplace
EOF
```

To check the status of the install you can run:

```
kubectl get installplan -n istio-system -o=jsonpath='{.items[0].status.phase}'
```

once ready it will be marked `complete` while installing it will be marked `installing`

#### Configure Istio

```
kubectl apply -f - <<EOF
apiVersion: operator.istio.io/v1alpha1
kind: Istio
metadata:
  annotations:
    argocd.argoproj.io/sync-wave: "1"
  name: default
spec:
  version: v1.21.0
  namespace: istio-system
  # Disable autoscaling to reduce dev resources
  values:
    pilot:
      autoscaleEnabled: false
EOF
```

Wait for Istio to be ready

```
kubectl wait istio/default -n istio-system --for="condition=Ready=true"
```

> ## TODO Thanos and Configure Observability

Author: @david-martin wondering what we want to put in here?

```
kubectl get installplan -n istio-system -o=jsonpath='{.items[0].status.phase}'
```

once ready it will be marked `complete` while installing it will be marked `installing`

#### Configure Istio

```
kubectl apply -f - <<EOF
apiVersion: operator.istio.io/v1alpha1
kind: Istio
metadata:
  annotations:
    argocd.argoproj.io/sync-wave: "1"
  name: default
spec:
  version: v1.21.0
  namespace: istio-system
  # Disable autoscaling to reduce dev resources
  values:
    pilot:
      autoscaleEnabled: false
EOF
```

Wait for Istio to be ready

```
kubectl wait istio/default -n istio-system --for="condition=Ready=true"
```

### (Recommended) Thanos and Observability Stack

Kuadrant provides a set of sample dashboards that use the known metrics exported by kuadrant and gateway components to provide insight into the different areas of your APIs and Gateways. While this is not essential it is recommended that you set up an observability stack. Below are links to the OpenShift docs on this and also a link to help with the setup of Thanos for metrics storage.

OpenShift supports a user facing monitoring stack. This can be cofigured and setup via their documentation

> https://docs.openshift.com/dedicated/observability/monitoring/configuring-the-monitoring-stack.html 

Reviewer 3: The documentation link points to the openshift dedicated. This expects that openshift has some resources created by default. E.g. `The user-workload-monitoring-config ConfigMap object exists. This object is created by default when the cluster is created.`

- If you have user workload monitoring enabled. We Recommend configuring  remote write to a central storage system such as Thanos: 

> - [Remote Write Config](https://docs.openshift.com/dedicated/observability/monitoring/configuring-the-monitoring-stack.html#configuring_remote_write_storage_configuring-the-monitoring-stack)

Reviewer 3: Link to openshift dedicated. Applies the same as above.

```

kubectl create ns kuadrant-system

```

AWS Credential for TLS verification

```
kubectl -n kuadrant-system create secret generic aws-credentials \
  --type=kuadrant.io/aws \
  --from-literal=AWS_ACCESS_KEY_ID=$AWS_ACCESS_KEY_ID \
  --from-literal=AWS_SECRET_ACCESS_KEY=$AWS_SECRET_ACCESS_KEY
````  

Redis Credential for shared counters

```
kubectl -n kuadrant-system create secret generic redis-config \
  --from-literal=URL=$REDIS_URL  
```  

```
kubectl create ns ingress-gateway
```

AWS Credential for managing DNS records

```
kubectl -n ingress-gateway create secret generic aws-credentials \
  --type=kuadrant.io/aws \
  --from-literal=AWS_ACCESS_KEY_ID=$AWS_ACCESS_KEY_ID \
  --from-literal=AWS_SECRET_ACCESS_KEY=$AWS_SECRET_ACCESS_KEY
```  

To install the Kuadrant operator:

**Optional**:
Setup the catalog. Note this step is only needed if you want to use the latest cutting edge.
```

kubectl apply -f - <<EOF
apiVersion: operators.coreos.com/v1alpha1
kind: CatalogSource
metadata:
  name: kuadrant-operator-catalog
  namespace: kuadrant-system
spec:
  sourceType: grpc

>   image: quay.io/kuadrant/kuadrant-operator-catalog:latest

Reviewer 5: `latest` here is intentional, correct? We don't want to point to any specific tag.

Author: yep

Reviewer 2: I would actually disagree with this, we should never point to the latest in the documentation. We should always use release images

Author: can remove from this doc specifically, but it is useful for upstream docs in my opinion as long as the user knows what they are getting is the latest possible build

  displayName: Kuadrant Operators
  publisher: grpc
  updateStrategy:
    registryPoll:
      interval: 45m
EOF
```      
Install the Kuadrant Operator
```
kubectl apply -f - <<EOF
apiVersion: operators.coreos.com/v1alpha1
kind: Subscription
metadata:
  annotations:

>     argocd.argoproj.io/sync-wave: "0"

Reviewer 2: Is this annotation needed?

Author: removed

  name: kuadrant-operator
  namespace: kuadrant-system
spec:
  channel: preview
  installPlanApproval: Automatic
  name: kuadrant-operator
  source: kuadrant-operator-catalog
  sourceNamespace: kuadrant-system
---
kind: OperatorGroup
apiVersion: operators.coreos.com/v1
metadata:
  name: kuadrant
  namespace: kuadrant-system
spec: 
  upgradeStrategy: Default 
EOF
```  

Wait for kuadrant operators to be installed

```
kubectl get installplan -n kuadrant-system -o=jsonpath='{.items[0].status.phase}'
```

This should return `complete`

#### Configure Kuadrant

```
kubectl apply -f - <<EOF
apiVersion: kuadrant.io/v1beta1
kind: Kuadrant
metadata:
  annotations:
    argocd.argoproj.io/sync-wave: "1"
  name: kuadrant
  namespace: kuadrant-system
spec:
  limitador:
    storage:

>       redis:

Reviewer 5: It is recommended to use  `.storage.redis-cached`  for better performance.
